pennylane/tape/tape.py

```diff
@@ -930,6 +936,268 @@ def set_parameters(self, params, trainable_only=True):
+
+        elif ret_type == qml.measurements.Sample:
+            shape = []
+            for shot_val in device.shot_vector:
+                for _ in range(shot_val.copies):
+                    shots = shot_val.shots
+                    if shots != 1:
+                        shape.append(tuple([shots, len(mps)]))
+                    else:
+                        shape.append((len(mps),))
+        return shape
+
+    def shape(self, device):
+        """Produces the output shape of the tape by inspecting its measurements
+        and the device used for execution.
+
+        .. note::
+
+            The computed shape is not stored because the output shape may be
+            dependent on the device used for execution.
+
+        **Example:**
+
+        .. code-block:: python
+
+            import pennylane as qml
+            from pennylane.gradients import param_shift
+            from pennylane import numpy as np
+
+            dev = qml.device("default.qubit", wires=2)
+            a = np.array([0.1, 0.2, 0.3])
+
+            def func(a):
+                qml.RY(a[0], wires=0)
+                qml.RX(a[1], wires=0)
+                qml.RY(a[2], wires=0)
+
+            with qml.tape.QuantumTape() as tape:
+                func(a)
+                qml.state()
+
+            # 1. Execute the tape
+            tape_execute_res = qml.execute(
+                [tape],
+                dev,
+                cache=None,
+                interface="autograd",
+                gradient_fn=param_shift
+            )
+
+            # Get the result of the first tape execution
+            tape_execute_res = tape_execute_res[0]
+
+        .. code-block:: pycon
+
+            >>> print(tape_execute_res.shape)
+            (1, 4)
+
+        There is a difference between the shape of results obtained by:
+
+        1. Using the low-level ``qml.execute`` pipeline for executing a ``QuantumTape``;
+        2. Evaluating a QNode.
+
+        The shape returned by this method corresponds to case 1. QNode
+        evaluation may include an additional squeezing step.
+
+        For example, for `qml.state` the result in each case would be (see
+        example below):
+
+        1. ``(1,4)``;
+        2. ``(4,)``
+
+        .. code-block:: python
+
+            # 2. Evaluate the QNode
+            @qml.qnode(dev,interface='autograd')
+            def circuit(a):
+                func(a)
+                return qml.state()
+
+            res_qnode = circuit(a)
+
+        .. code-block:: pycon
+
+            >>> print(tape_execute_res.shape, res_qnode.shape)
+            (1, 4) (4,)
+
+        Args:
+            device (.Device): the device that will be used for the tape execution
+
+        Raises:
+            TapeError: raised for unsupported cases for
+                example when the tape contains heterogeneous measurements
+
+        Returns:
+            Union[tuple[int], list[tuple[int]]]: the output shape(s) of the
+            tape result
```

Comment:
Don't forget to put the example _below_ the arguments/returns!

Reply:
Updated in https://github.com/PennyLaneAI/pennylane/pull/2044/commits/7e4f99e962dab0459125f33a3efea1bb713d3f0a